### PR TITLE
#70 wrapper request aka requestentity

### LINKED ISSUE #70: Wrapper Request aka RequestEntity
Implement `BringRequest` based on [Spring RequestEntity](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/RequestEntity.html)

Implementation should include:
1. Parameterized body. Getter and setter. Parameterized boyd should handle `Void` type, for example if `BringRequest` will be used in `GET` request with no body.
2. Headers map. Map type `<String, String>`.
3. Cookies map binded to headers map. Map type `<String, String>`.
4. Http method type. Enumerate type describing what type of request method was used (GET, POST ...).
5. URL. URL of type `URI` containing endpoint full URL.
7. Builder for easy creation.

### PR DESCRIPTION
**Pull request by AlexandrKarpachov**

### CODE REVIEW

File: src/main/java/io/github/bobocodebreskul/context/exception/RequestsMappingException.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package io.github.bobocodebreskul.context.exception;
+
+public class RequestsMappingException extends RuntimeException {
```
Comment (RomanPryshchepa) [inline src/main/java/io/github/bobocodebreskul/context/exception/RequestsMappingException.java:10]: Documentation is absent

File: src/main/java/io/github/bobocodebreskul/server/HttpRequestMapper.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+  public HttpRequestMapper() {
+
+  }
+
+  public HttpRequestMapper(ObjectMapper objectMapper) {
+    this.objectMapper = objectMapper;
+  }
+
+  // todo test BringResponse with all fields then we set correct values to HttpServletResponse
```
Comment (RomanPryshchepa) [inline src/main/java/io/github/bobocodebreskul/server/HttpRequestMapper.java]: Remove todo

File: src/main/java/io/github/bobocodebreskul/server/HttpRequestMapper.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        httpServletResponse.setHeader(CONTENT_TYPE_KEY, contentType);
+      }
+    } catch (IOException e) {
+      log.error("Failed to write response entity to httpServletResponse", e);
+      throw new RequestsMappingException("Failed to write response entity to httpServletResponse",
+          e);
+    }
+  }
+
+  // todo test when we fill all BringRequest fields from HttpServletRequest when body exists
```
Comment (RomanPryshchepa) [inline src/main/java/io/github/bobocodebreskul/server/HttpRequestMapper.java]: Remove todo

File: src/main/java/io/github/bobocodebreskul/server/TomcatServer.java [hunk 4]
```diff
@@ -48,7 +48,7 @@ public class TomcatServer {
    */
   public static void run(BringContainer container) {
     log.info("Tomcat server is starting...");
-    disableTomcatLogs();
+//    disableTomcatLogs();
```
Comment (RomanPryshchepa) [inline src/main/java/io/github/bobocodebreskul/server/TomcatServer.java]: What is it?

File: src/main/java/io/github/bobocodebreskul/server/BringHttpEntity.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    if (this.headers.containsKey(lowerCaseHeaderName)) {
+      String headerValue = this.headers.get(lowerCaseHeaderName);
+      log.info("Header value for header name '{}' is '{}'.", headerName, headerValue);
+      return headerValue;
+    }
+    log.error("Header for name='{}' not found.", headerName);
+    return "";
+  }
+
+  public Set<String> getHeadersNames() {
```
Comment (s-barabash) [inline src/main/java/io/github/bobocodebreskul/server/BringHttpEntity.java:87]: add java docs

**Review by mykolaFilimonov — APPROVED** [review]

**Review by RomanPryshchepa — APPROVED** [review]

File: src/main/java/io/github/bobocodebreskul/server/BringHttpEntity.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+   */
+  public String getHeader(String headerName) {
+    String lowerCaseHeaderName = headerName.toLowerCase();
+    log.info("Get header for headerName='{}'.", headerName);
+    if (this.headers.containsKey(lowerCaseHeaderName)) {
+      String headerValue = this.headers.get(lowerCaseHeaderName);
+      log.info("Header value for header name '{}' is '{}'.", headerName, headerValue);
+      return headerValue;
+    }
+    log.error("Header for name='{}' not found.", headerName);
```
Comment (s-barabash) [inline src/main/java/io/github/bobocodebreskul/server/BringHttpEntity.java]: it should be warn not error

File: src/main/java/io/github/bobocodebreskul/server/BringHttpEntity.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+  }
+
+  private String getCookiesAsString() {
+    List<String> pairs = this.cookies.entrySet().stream()
+        .map(entry -> entry.getKey() + "=" + entry.getValue())
+        .toList();
+    return String.join("; ", pairs);
+  }
+
+  public T getBody() {
```
Comment (s-barabash) [inline src/main/java/io/github/bobocodebreskul/server/BringHttpEntity.java:190]: add java docs

File: src/main/java/io/github/bobocodebreskul/server/BringHttpEntity.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    log.debug("Get body call");
+    return body;
+  }
+
+  public void setBody(T body) {
```
Comment (s-barabash) [inline src/main/java/io/github/bobocodebreskul/server/BringHttpEntity.java:199]: add java docs

File: src/main/java/io/github/bobocodebreskul/server/BringRequest.java [hunk 9]
```diff
@@ -1,5 +1,102 @@
 package io.github.bobocodebreskul.server;
 
-public class BringRequest {
+import io.github.bobocodebreskul.server.enums.RequestMethod;
+import java.net.URI;
+import java.util.Map;
+import lombok.extern.slf4j.Slf4j;
 
+// TODO: tests
```
Comment (s-barabash) [inline src/main/java/io/github/bobocodebreskul/server/BringRequest.java]: remove todo

File: src/main/java/io/github/bobocodebreskul/server/BringRequest.java [hunk 10]
```diff
@@ -1,5 +1,102 @@
...
+    private final RequestMethod requestMethod;
+    private final URI uri;
+    private Map<String, String> headers;
+
+    private Builder(RequestMethod requestMethod, URI uri) {
+      this.requestMethod = requestMethod;
+      this.uri = uri;
+    }
+
+    public Builder headers(Map<String, String> headers) {
```
Comment (s-barabash) [inline src/main/java/io/github/bobocodebreskul/server/BringRequest.java:86]: add java docs

File: src/main/java/io/github/bobocodebreskul/server/BringRequest.java [hunk 11]
```diff
@@ -1,5 +1,102 @@
...
+      this.headers = headers;
+      return this;
+    }
+
+    public <T> BringRequest<T> body(T body) {
```
Comment (s-barabash) [inline src/main/java/io/github/bobocodebreskul/server/BringRequest.java:98]: add java docs

File: src/main/java/io/github/bobocodebreskul/server/BringRequest.java [hunk 12]
```diff
@@ -1,5 +1,102 @@
...
+      return create(body);
+    }
+
+    public BringRequest<Void> build() {
```
Comment (s-barabash) [inline src/main/java/io/github/bobocodebreskul/server/BringRequest.java:107]: add java docs

File: src/main/java/io/github/bobocodebreskul/server/BringResponse.java [hunk 13]
```diff
@@ -28,34 +26,23 @@
...
-  static final String COOKIE_VALUE_SHOULD_NOT_BE_NULL = "Cookie value should not be null.";
+  private static final Logger log = LoggerFactory.getLogger(BringResponse.class);
 
-  private Map<String, String> headers = new HashMap<>();
-
-  private final Map<String, String> cookies = new HashMap<>();
-  private T body;
   private ResponseStatus status;
 
   public BringResponse(T body) {
```
Comment (s-barabash) [inline src/main/java/io/github/bobocodebreskul/server/BringResponse.java:35]: add java docs

**Review by s-barabash — APPROVED** [review]

**Review by JoshukLite — APPROVED** [review]

**Pull request merged**